src/wrapper/jnienv.rs

```diff
@@ -1932,6 +1967,202 @@ impl<'a> JNIEnv<'a> {
+        jni_void_call!(
+            self.internal,
+            ReleaseByteArrayElements,
+            array,
+            elems,
+            jni_sys::JNI_COMMIT
+        );
+        Ok(())
+    }
+
+    /// Return an AutoByteArray of the given Java byte array.
+    ///
+    /// The result is valid until the AutoByteArray object goes out of scope, when the
+    /// release happens automatically according to the mode parameter.
+    ///
+    /// Since the returned array may be a copy of the Java array, changes made to the
+    /// returned array will not necessarily be reflected in the original array until
+    /// release_array_elements() is called.
+    /// AutoByteArray has a commit() method, to force a copy of the array if needed (and without
+    /// releasing it).
+    ///
+    /// See also [`get_byte_array_elements`](struct.JNIEnv.html#method.get_byte_array_elements)
+    pub fn get_byte_array_elements_auto(
+        &self,
+        array: jbyteArray,
+        mode: ReleaseMode,
+    ) -> Result<AutoByteArray> {
+        let mut is_copy: jboolean = 0xff;
+        let ptr = self.get_byte_array_elements(array, &mut is_copy).unwrap();
+        let res = self.auto_byte_array(array, ptr, mode, is_copy != JNI_FALSE);
+        Ok(res)
+    }
+
+    /// Return a pointer to elements of the given Java primitive array.
+    ///
+    /// The semantics of this function are very similar to the byte_array_elements function. If
+    /// possible, the VM returns a pointer to the primitive array; otherwise, a copy is made.
+    /// However, there are significant restrictions on how these functions can be used.
+    ///
+    /// After calling get_primitive_array_critical, the native code should not run for an extended
+    /// period of time before it calls release_primitive_array_critical. We must treat the code
+    /// inside this pair of functions as running in a "critical region." Inside a critical region,
+    /// native code must not call other JNI functions, or any system call that may cause the
+    /// current thread to block and wait for another Java thread. (For example, the current
+    /// thread must not call read on a stream being written by another Java thread.)
+    ///
+    /// These restrictions make it more likely that the native code will obtain an uncopied version
+    /// of the array, even if the VM does not support pinning. For example, a VM may temporarily
+    /// disable garbage collection when the native code is holding a pointer to an array obtained
+    /// via get_primitive_array_critical.
+    ///
+    /// Note that get_primitive_array_critical might still make a copy of the array if the VM
+    /// internally represents arrays in a different format. Therefore, we need to check its
+    /// return value against NULL for possible out of memory situations.
+    pub fn get_primitive_array_critical(
+        &self,
+        array: jarray,
+        is_copy: &mut jboolean,
```

Comment:
A tuple / struct doesn't make much sense, as the function is returning only one thing. In fact, if you're OK with that I can return the raw pointer directly, instead of wrapping it into a `Result`. If I recall correctly, I only wrapped it because that's what I saw in other functions / examples.

Reply:
Update: Or are you're talking about removing the `is_copy` parameter entirely, and returning a tuple / struct with the pointer and the `is_copy` flag? I'm OK with that.

Reply:
Yes, exactly. I think that "output parameters" are confusing, because in that case it is non-obvious if it is out or in-out parameter.

Reply:
Changed.

As part of this I also added a non-null check to `get_primitive_array_critical`'s return value, which was missing.

Also changed / fixed the documentation.

Thanks for your review. The code is now cleaner, and more elegant.